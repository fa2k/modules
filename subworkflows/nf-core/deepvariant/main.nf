
include { DEEPVARIANT_MAKEEXAMPLES        } from '../../../modules/nf-core/deepvariant/makeexamples/main'
include { DEEPVARIANT_CALLVARIANTS        } from '../../../modules/nf-core/deepvariant/callvariants/main'
include { DEEPVARIANT_POSTPROCESSVARIANTS } from '../../../modules/nf-core/deepvariant/postprocessvariants/main'

workflow DEEPVARIANT {
    take:
    ch_input            // channel: [ val(meta), path(input), path(index), path(intervals)]
    ch_fasta            // channel: [ val(meta2), path(fasta) ]
    ch_fai              // channel: [ val(meta3), path(fail) ]
    ch_gzi              // channel: [ val(meta4), path(gzi) ]
    ch_model_type       // channel: val("wgs" | "wes")

    main:

    ch_versions = Channel.empty()

    DEEPVARIANT_MAKEEXAMPLES(ch_input, ch_fasta, ch_fai, ch_gzi)
    ch_versions = ch_versions.mix(DEEPVARIANT_MAKEEXAMPLES.out.versions.first())

    DEEPVARIANT_CALLVARIANTS(DEEPVARIANT_MAKEEXAMPLES.out.examples, ch_model_type)
    ch_versions = ch_versions.mix(DEEPVARIANT_CALLVARIANTS.out.versions.first())
    
    // Input to postprocessing step needs both the gvcfs from MAKEEXAMPLES and the variant
    // calls from CALLVARIANTS. Joining on the tuple element make_examples_id.
    ch_postproc_input = DEEPVARIANT_CALLVARIANTS.out.call_variants_tfrecords.join(
        DEEPVARIANT_MAKEEXAMPLES.out.gvcf,
        by: [0, 1], // Join on [0] is sufficient, but we want to collapse meta as well.
        failOnMismatch: true
    ).map { it.drop(1) } // Drop the unique ID, which shouldn't be input into POSTPROCESSVARIANTS
    DEEPVARIANT_POSTPROCESSVARIANTS(
        ch_postproc_input,
        ch_fasta,
        ch_fai,
        ch_gzi
    )

    ch_versions = ch_versions.mix(DEEPVARIANT_POSTPROCESSVARIANTS.out.versions.first())

    emit:
    vcf         = DEEPVARIANT_POSTPROCESSVARIANTS.out.vcf
    vcf_tbi     = DEEPVARIANT_POSTPROCESSVARIANTS.out.vcf_tbi
    gvcf        = DEEPVARIANT_POSTPROCESSVARIANTS.out.gvcf
    gvcf_tbi    = DEEPVARIANT_POSTPROCESSVARIANTS.out.gvcf_tbi
    versions    = ch_versions
}
