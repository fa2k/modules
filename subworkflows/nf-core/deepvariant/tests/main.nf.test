nextflow_workflow {

    name "Test Subworkflow DEEPVARIANT"
    script "../main.nf"
    workflow "DEEPVARIANT"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/deepvariant"

    tag "deepvariant"
    tag "deepvariant/makeexamples"
    tag "deepvariant/callvariants"
    tag "deepvariant/postprocessvariants"

    test("homo_sapiens - [bam, bai] - fasta - fai") {
        when {

            workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                      []
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                input[4] = [
                    "wgs"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
/*
    test("homo_sapiens - [cram, crai, genome_bed] - fasta - fai") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram.crai', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.bed', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("homo_sapiens - [bam, bai] - fasta_gz - fasta_gz_fai") {
        when {

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                      []
                ]
                input[1] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz.fai', checkIfExists: true)
                ]
                input[3] = [
                    [ id:'genome'],
                      file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.gz.gzi', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
*/
}
