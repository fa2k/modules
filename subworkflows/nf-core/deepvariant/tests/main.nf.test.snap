{
    "homo_sapiens - different samples and regions - cram - fasta - fai": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz:md5,1a1d1d4e0cee5dea34989029185e145c"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz.tbi:md5,6f6a5648676395fb8256f26df6a46ae1"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz:md5,65d6b5a3a8cfcde6b6a63bee8e6e7982"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz.tbi:md5,62f45c757d51f2449cff6a38cec61192"
                    ]
                ],
                "4": [
                    "versions.yml:md5,328e2f6462a2b9a5b4c291d2ea2acbae",
                    "versions.yml:md5,3ee8fb91cc070d641058839976671aa3",
                    "versions.yml:md5,c4510740784482ca098220df79ba3f3e"
                ],
                "gvcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz:md5,65d6b5a3a8cfcde6b6a63bee8e6e7982"
                    ]
                ],
                "gvcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz.tbi:md5,62f45c757d51f2449cff6a38cec61192"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz:md5,1a1d1d4e0cee5dea34989029185e145c"
                    ]
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz.tbi:md5,6f6a5648676395fb8256f26df6a46ae1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,328e2f6462a2b9a5b4c291d2ea2acbae",
                    "versions.yml:md5,3ee8fb91cc070d641058839976671aa3",
                    "versions.yml:md5,c4510740784482ca098220df79ba3f3e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-08T11:06:33.690696159"
    },
    "homo_sapiens - two inputs - bam - fasta - fai": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ]
                ],
                "4": [
                    "versions.yml:md5,328e2f6462a2b9a5b4c291d2ea2acbae",
                    "versions.yml:md5,3ee8fb91cc070d641058839976671aa3",
                    "versions.yml:md5,c4510740784482ca098220df79ba3f3e"
                ],
                "gvcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz:md5,f54f28e0cede88517a55ed3bd832fd00"
                    ]
                ],
                "gvcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.g.vcf.gz.tbi:md5,a727c8b9a4a98bb22a2946e7b729d65d"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz:md5,1cd3ae45139543b48a5f9a8557fa476a"
                    ]
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ],
                    [
                        {
                            "id": "test2",
                            "single_end": false
                        },
                        "test2.vcf.gz.tbi:md5,d0f3cbb281459620d545581586c4cb89"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,328e2f6462a2b9a5b4c291d2ea2acbae",
                    "versions.yml:md5,3ee8fb91cc070d641058839976671aa3",
                    "versions.yml:md5,c4510740784482ca098220df79ba3f3e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-08T11:05:58.39494171"
    }
}